pkg/kubelet/kuberuntime/kuberuntime_manager.go

```diff
@@ -494,6 +507,17 @@ func (m *kubeGenericRuntimeManager) computePodActions(pod *v1.Pod, podStatus *ku
 			changes.NextInitContainerToStart = &pod.Spec.InitContainers[0]
 			return changes
 		}
+
+		if utilfeature.DefaultFeatureGate.Enabled(features.SidecarLifecycle) {
```

Comment:
if there is init container, seems like this code path will never be reached as sidecars will never get created

Reply:
Also, seems a pretty weak assertion that we wait for sidecar to be “ready” once before creating standard containers. Coz a container can be in CrashLoopBackoff so it’s technically impossible that we create standard container after “all sidecars are ready”

 I’d still prefer we don’t keep such order but if people already agreed on it in design doc, I’m ok with proceeding with it.

Reply:
This was considered essential for this feature to be useful, if we don't wait for the sidecars to be ready before starting the non-sidecars, what's the advantage of starting them first at all?

Reply:
They question about “wait for the sidecars to be ready” is, are we holding the assumption that sidecar does not fall in CrashLoopBackoff? As from code we only wait sidecars to be ready *once*

Also from current implcode, if there are init containers, sidecars will not be created at all

Reply:
Sidecars are created if there are init containers, they are created after the init containers have completed, but before the non-sidecars are started, I have tested this to verify, I will add some unit tests as well. That code path is only for when creating the pod sandbox.

So the current implementation waits for the sidecars to all become ready before starting the non-sidecars, if the side-car starts up and immediately crashloops then your non-sidecars would never be started. If your side-car becomes ready and then crashloops , your non-sidecars will have already started and the pod would behave as normal.

Reply:
I might have missed a bit here, the logic is like, 
if need to create pod sand box, the if block gets first sidecar and return, and next `computePodAction` call will not enter `if createPodSandbox {}` block anymore

aka, if there are init containers, your create sidecar code block will not be touched here, and standard container gets actually created at https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L567,

and such case does not seem to be handled here

Reply:
I think @zhan849 is suggesting we bring the sidecar ContainersToStart field out of the `if createPodSandbox` block. The order of events is:

- pod is fresh, no sandbox exists
- if createPodSandbox is true, so we start the first init container and early return
- we go back into the loop, but the sandbox is already made so we skip that block

As a result, sidecars are never started first, we just start all of them at once.

Reply:
@tomwans @Joseph-Irving 

so the current logic is a little bit redundant, it has special handlings of scenarios with/without init container in `if createPodSandbox ` block. (I actually tried to merge logics but this function just has too many corner cases and can easily be broken)

I'd suggest to follow the current convention:
```go
if createSandbox {
  if hasInit {
    return first init
  } else if hasSidecar {
    return sidecars
  } else {
    return standards
  }
}
if hasInit {
  return firstInit
}
if hasSidecar {
  return sidecars
}

if sidecar not ready {
  return
}

startNormalContainers

......


```

Reply:
To summarise the current logic:
```golang
if createSandbox {
  if hasInit {
    return first init
  } else if hasSidecar {
    return sidecars
  } else {
    return containers
  }
}
if initalizing {
  return nextInit
}

for containers {
  ifSidecarsNotReady && container != sidecar {
    continue
  } else {
  compute actions on container, e.g start the container, restart it if failed etc
 }
}
```
So once the init containers have finished we can start iterating over the containers. The non-sidecars are ignored if the sidecars haven't started and became ready. So this loops sends back the sidecars in containersToStart. If they have all become ready, the next time ComputePodActions is called all the non-sidecar containers will be sent back in ContainersToStart. Note this check of whether the sidecars are ready is ignored after the normal containers have started, this is so that we only block the normal containers from starting during startup.

Reply:
to follow up on this - if pod has init, you will first hit
```go
if createSandbox {
  if hasInit {
    return first init   // <----
  } else if hasSidecar {
    return sidecars
  } else {
    return containers
  }
}
```
next time when we hit `computePodActions()` we will no longer hit `if createSandbox` block, so where is the place we start all sidecars?

Reply:
we start them here
```
for containers {
  ifSidecarsNotReady && container != sidecar {
    continue
  } else {
  compute actions on container, e.g start the container, restart it if failed etc
 }
}
```
Just like you would start the normal containers, they get started using the same loop

Reply:
I've already responded to this @zhan849, maybe you missed this, there is also a test showing this behaviour https://github.com/kubernetes/kubernetes/pull/80744/files#diff-626ea7fe1e5c7fe6e6f4cbd59345b44aR1613-R1641